### PR TITLE
feat: Derive Eq & Default

### LINKED ISSUE #4594: Epic: Metaprogramming
Proposal: https://hackmd.io/Tkzo_ryvTsWMPESmWHPiZw?view

```[tasklist]
### Core Tasks for Noir 1.0
- [ ] https://github.com/noir-lang/noir/issues/4586
- [ ] https://github.com/noir-lang/noir/issues/4587
- [ ] https://github.com/noir-lang/noir/issues/4588
- [ ] https://github.com/noir-lang/noir/issues/4589
- [ ] https://github.com/noir-lang/noir/issues/4590
- [ ] https://github.com/noir-lang/noir/issues/4591
- [ ] https://github.com/noir-lang/noir/issues/4796
- [ ] https://github.com/noir-lang/noir/issues/4916
- [ ] https://github.com/noir-lang/noir/issues/4917
- [ ] https://github.com/noir-lang/noir/issues/4924
- [ ] https://github.com/noir-lang/noir/issues/4925
- [ ] https://github.com/noir-lang/noir/issues/4953
- [ ] https://github.com/noir-lang/noir/issues/4954
- [ ] https://github.com/noir-lang/noir/issues/5146
- [ ] https://github.com/noir-lang/noir/issues/5168
- [ ] https://github.com/noir-lang/noir/issues/5201
- [ ] https://github.com/noir-lang/noir/issues/5255
- [ ] https://github.com/noir-lang/noir/issues/5285
- [ ] https://github.com/noir-lang/noir/issues/5284
- [ ] https://github.com/noir-lang/noir/issues/5473
- [ ] https://github.com/noir-lang/noir/issues/5476
- [ ] https://github.com/noir-lang/noir/issues/5477
- [ ] https://github.com/noir-lang/noir/issues/5480
- [ ] https://github.com/noir-lang/noir/issues/5481
```

```[tasklist]
### Bugs for Noir 1.0
- [ ] https://github.com/noir-lang/noir/issues/4879
- [ ] https://github.com/noir-lang/noir/issues/4915
- [ ] https://github.com/noir-lang/noir/issues/5194
- [ ] https://github.com/noir-lang/noir/issues/5333
- [ ] https://github.com/noir-lang/noir/issues/5388
- [ ] https://github.com/noir-lang/noir/issues/5482
- [ ] https://github.com/noir-lang/noir/issues/5575
```

```[tasklist]
### Nice-to-haves for Noir 1.0+
- [ ] https://github.com/noir-lang/noir/issues/3379
- [ ] https://github.com/noir-lang/noir/issues/4922
- [ ] https://github.com/noir-lang/noir/issues/4592
- [ ] https://github.com/noir-lang/noir/issues/5478
- [ ] https://github.com/noir-lang/noir/issues/5479
```

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves #4594 

First traits to derive in the stdlib!

## Summary\*

Lets us derive the `Eq` and `Default` traits for almost any type.

I anticipate there will be issues with deriving generic types. I'll fix this in another PR.

## Additional Context

Since derive is in the stdlib and we can derive traits now, I'm arbitrarily marking the epic for metaprogramming as completed. There are still bugs to fix and many helper functions we can add but the bulk of the work and infrastructure is all in.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [x] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: noir_stdlib/src/cmp.nr [hunk 1]
```diff
@@ -1,9 +1,37 @@
...
+
+    let where_clause = s.generics().map(|name| quote { $name: Default }).join(quote {,});
+
+    // `(self.a == other.a) & (self.b == other.b) & ...`
+    let equalities = s.fields().map(
+        |f: (Quoted, Type)| {
+        let name = f.0;
+        quote { (self.$name == other.$name) }
+    }
+    );
```
Comment (jfecher) [inline noir_stdlib/src/cmp.nr:23]: Nargo fmt is a bit gnarly here

File: noir_stdlib/src/hash/poseidon2.nr [hunk 2]
```diff
@@ -1,7 +1,7 @@
 use crate::hash::Hasher;
 use crate::default::Default;
 
-global RATE: u32 = 3;
+comptime global RATE: u32 = 3;
```
Comment (jfecher) [inline noir_stdlib/src/hash/poseidon2.nr:4]: Needed to mark this as global to be used in the comptime hashing code.

It's a bit awkward to do this automatically since it means lazily evaluating globals in the interpreter which aren't marked comptime. This could change program behavior if the evaluating the global has side-effects. So I opted to just mark this as comptime for now.

File: noir_stdlib/src/meta/mod.nr [hunk 3]
```diff
@@ -2,9 +2,9 @@ use crate::collections::umap::UHashMap;
 use crate::hash::BuildHasherDefault;
 use crate::hash::poseidon2::Poseidon2Hasher;
 
+mod struct_def;
```
Comment (jfecher) [inline noir_stdlib/src/meta/mod.nr:5]: Unrelated change (sorry): Renamed the type_def module to struct_def now that the `TypeDefinition` type has been renamed to `StructDefinition`.

**Review by vezenovm — APPROVED** [review]
Lets goo!
